Native SPI RGB LED component

clydebarrow opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

This new component supports the APA102 and similar SPI addressable RGB LEDs. While there is support for these LEDs in other components, these rely on external libraries supporting multiple different chip types and don't work with ESP-IDF.

This component supports only the SPI (two wire) LED interface and leverages the existing ESPHome `spi` component so works on both Arduino and ESP-IDF platforms, is lightweight and performant. It has only been tested on ESP32 but should work with any platform supported by ESPHome that implements SPI.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3147
## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
    spi:
      mosi_pin: GPIO3
      clk_pin: GPIO9

    light:
      - platform: apa102
        num_leds: 30
        color_correct: [80%, 60%, 100%]
        id: rgb_led
        name: "RGB LED Strip"
        data_rate: 1MHz
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

clydebarrow committed:
Add testing branch to workflow

clydebarrow committed:
Add workflow

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into dev

clydebarrow committed:
Checkpoint

clydebarrow committed:
Align SPI data rates in c++ code with Python code.

clydebarrow committed:
Merge branch 'spirates' into apa102

clydebarrow committed:
Checkpoint

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into apa102

clydebarrow committed:
CI fixes

clydebarrow committed:
Update codeowners

clydebarrow committed:
Workflow cleanup

jesserockz commented:
This is kinda a duplicate of #5101 which does need work to use SPIDevice like this PR does, but implements a generic SPI led strip driver which would be preferred over having multiple specific ones like this PR implements. Perhaps somehow a merge of the two PRs is in order?

clydebarrow commented:
> generic SPI led strip driver which would be preferred over having multiple specific ones like this PR implements

Well I don't have any of the other LEDs on hand to test with, however looking at the datasheet for the SK9822, APA107 and the HD107 it appears the code I already wrote will work with them, as they have the identical frame format and BGR order. So maybe all that's needed is a name change (maybe spi_rgb_led) and some extra documentation. 

I don't want to make it some kind of umbrella component that tries to do too much. Keeping it simple makes keeping it working easy. IOW I would resist adding the one-wire LED protocol support which requires quite different code.

clydebarrow commented:
The APA107 is a 1-wire chip, so not SPI compatible. The HD108 is SPI but has 16 bit data, so that's a possibility to add, but I'd prefer to do that later when I have some to test with. They're new and more expensive so probably not widely in use yet.

So right now it seems the APA102, APA102C, HD107 and SK9822 will work with my code and they are the main SPI type RGB leds available.

clydebarrow committed:
Rename to spi_rgb_led

clydebarrow committed:
Rename header file

clydebarrow committed:
Clang tidy

clydebarrow committed:
Merge branch 'dev' into apa102

kbx81 commented:
> The APA107 is a 1-wire chip

FWIW, I think the APA107 is also a 2-wire interface with clock and data lines -- at least, everything I'm finding on it right now seems to indicate this. 😇

clydebarrow commented:
I found two conflicting datasheets for APA107. Maybe there are two different chips from different manufacturers.

https://www.rose-lighting.com/wp-content/uploads/sites/53/2019/10/apa107-LED.pdf

https://www.superlightingled.com/PDF/APA107.pdf

clydebarrow commented:
Anyway, the 2 wire APA107 should work just fine with this component. I renamed it to `spi_rgb_led` and updated the docs to reflect the fact it works with chips other than APA102.

clydebarrow committed:
Disable spi after transfer.

clydebarrow committed:
Merge branch 'apa102' of https://github.com/clydebarrow/esphome into apa102

clydebarrow committed:
Move enable() to where it belongs

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into apa102

clydebarrow committed:
Call spi_setup before enable

clydebarrow committed:
Clang tidy

clydebarrow committed:
Add test

clydebarrow committed:
Merge branch 'dev' into apa102

clydebarrow committed:
Merge branch 'dev' into apa102

jesserockz commented:
`spi_led_strip` would be better to match `(esp32_)rmt_led_strip` and `rp2040_pio_led_strip`

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into apa102

clydebarrow committed:
Rename to spi_led_strip

kbx81 commented on esphome/components/spi_led_strip/spi_led_strip.h:
```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include "esphome/core/component.h"
```
I believe you need to add `defines.h` to ensure `SPIRGBLED_DATA_RATE` is picked up. [See also.](https://github.com/esphome/esphome/pull/5344)

```suggestion
#include "esphome/core/component.h"
#include "esphome/core/defines.h"
```

clydebarrow replied:
Is `defines.h` something new? Because it worked as it was.

But now that you have merged the new SPI PR, that won't be necessary. I'll update it to dynamically set the rate rather than using the #define kludge.

kbx81 replied:
> Is `defines.h` something new? Because it worked as it was.

It'll likely work because many other components include it; it's where any defines added to the project are written by `cg.add_define`. If by some chance nothing includes it, the macro wouldn't work. 😇

> But now that you have merged the new SPI PR, that won't be necessary. I'll update it to dynamically set the rate rather than using the #define kludge.

Sounds great! I'll leave you to it, then... 😁

kbx81 replied:
Also, feel free to nuke my last commits -- was just trying to resolve a merge conflict and I fudged up the codeowners. 😅

clydebarrow replied:
Ok pending the CI run, should be good to go. Tested here with the changes.

kbx81 replied:
Awesome! I'll give it another test on my FunHouse board 😄

kbx81 committed:
Include 'defines.h'

kbx81 committed:
Merge branch 'dev' into apa102

kbx81 committed:
Fix CODEOWNERS

clydebarrow committed:
Migrate data rate to new style setting.

clydebarrow committed:
Remove defines.h

clydebarrow committed:
Fix class name

clydebarrow committed:
Fix name in .py

clydebarrow committed:
And more more name tidy up.
